### PR TITLE
[MRG] Projection Robust Wasserstein

### PR DESCRIPTION
**Pull request by mhhuang95**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and context / Related issue

Code for
A Riemannian Block Coordinate Descent Method for Computing the PRW Distance, ICML 2021
Source: https://github.com/mhhuang95/PRW_RBCD

## How has this been tested (if it applies)

tested on Fragmented Hypercube problem

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](CONTRIBUTING.md) document.
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by agramfort** [comment]
You have failing tests @mhhuang95

File: ot/dr.py [hunk 1]
```diff
@@ -198,3 +199,118 @@ def proj(X):
         return (X - mx.reshape((1, -1))).dot(Popt)
 
     return Popt, proj
+
+
+def prw(X, Y, a, b, tau, U0=None, reg=0.1, k=2, stopThr=1e-3, maxiter=100, verbose=0):
+    r"""
+    Projection Robust Wasserstein Distance _[12],[13]
```
Comment (agramfort) [inline ot/dr.py:206]: I doubt this is formatted properly to have links to references below

File: ot/dr.py [hunk 2]
```diff
@@ -198,3 +199,118 @@ def proj(X):
...
+    - :math:`H(\pi)` is entropy regularizer
+    - :math:`x_i`, `y_j` are samples of measures \mu and \nu respectively
+    
+    Parameters
+    ----------
+    X : ndarray, shape (n, d)
+        Samples from measure \mu
+    Y : ndarray, shape (n, d)
+        Samples from measure \nu
+    a : ndarray, shape (n, 1)
```
Comment (agramfort) [inline ot/dr.py:223]: Is it necessary to have ndim == 2 with shape[1] == 1? I would favor a flat vector for a and b

File: ot/dr.py [hunk 3]
```diff
@@ -198,3 +199,118 @@ def proj(X):
...
+    tau : float
+        stepsize for Riemannian Gradient Descent
+    U0 : ndarray, shape (d, p)
+        Initial starting point for projection.
+    reg : float, optional
+        Regularization term >0 (entropic regularization)
+    k : int
+        Subspace dimension
+    stopThr : float, optional
+        Accuracy
```
Comment (agramfort) [inline ot/dr.py:236]: Accuracy is to me misleading. Maybe tolerance ? But on what criteria?

File: ot/dr.py [hunk 4]
```diff
@@ -198,3 +199,118 @@ def proj(X):
...
+    v = np.ones(m) / m
+    ones = np.ones((n, m))
+
+    assert d > k
+
+    if U0 is None:
+        U = np.random.randn(d, k)
+        U, _ = np.linalg.qr(U)
+    else:
+        U = U0
```
Comment (agramfort) [inline ot/dr.py:271]: This line need to be covered by a test

File: ot/dr.py [hunk 5]
```diff
@@ -198,3 +199,118 @@ def proj(X):
         return (X - mx.reshape((1, -1))).dot(Popt)
 
     return Popt, proj
+
+
+def prw(X, Y, a, b, tau, U0=None, reg=0.1, k=2, stopThr=1e-3, maxiter=100, verbose=0):
```
[the remark this answers is not in the document]
Comment (rflamary) [reply ot/dr.py:204]: I agree projection_robust_wasserstein is longer but explicit, we will also need to change the wda and pca functions and add a deprecation on the old names.

File: ot/dr.py [hunk 6]
```diff
@@ -198,3 +199,116 @@ def proj(X):
...
+
+
+def projection_robust_wasserstein(X, Y, a, b, tau, U0=None, reg=0.1, k=2, stopThr=1e-3, maxiter=100, verbose=0):
+    r"""
+    Projection Robust Wasserstein Distance [32]
+
+    The function solves the following optimization problem:
+
+    .. math::
+        max_{U \in St(d, k)} min_{\pi \in \Pi(\mu,\nu)} \sum_{i,j} \pi_{i,j}*||U^T(x_i - y_j)||^2 - reg * H(\pi)
```
Comment (agramfort) [inline ot/dr.py:211]:
```suggestion
        \max_{U \in St(d, k)} \min_{\pi \in \Pi(\mu,\nu)} \sum_{i,j} \pi_{i,j} \|U^T(x_i - y_j)\|^2 - reg * H(\pi)
```

File: ot/dr.py [hunk 7]
```diff
@@ -198,3 +199,116 @@ def proj(X):
...
+    
+    - :math:`U` is a linear projection operator in the Stiefel(d, k) manifold
+    - :math:`H(\pi)` is entropy regularizer
+    - :math:`x_i`, `y_j` are samples of measures \mu and \nu respectively
```
Comment (agramfort) [inline ot/dr.py:215]:
```suggestion
    - :math:`x_i`, :math:`y_j` are samples of measures \mu and \nu respectively
```

**Review by agramfort — APPROVED** [review]

**Comment by rflamary** [comment]
Thank you again @mhhuang95 for this contribution and welcome to the POT contributors.

**Pull request merged**